libvast/src/concept/parseable/vast/type.cpp

```diff
@@ -112,17 +112,56 @@ bool type_parser::parse(Iterator& f, const Iterator& l, Attribute& a) const {
   static auto placeholder_parser
     = (parsers::identifier) ->* to_named_none_type
     ;
+  rule<Iterator, type> type_expr_parser;
+  auto algebra_leaf_parser
+    = record_type_parser
+    | placeholder_parser
+    ;
+  auto algebra_operand_parser
+    = algebra_leaf_parser
+    | ref(type_expr_parser)
+    ;
+  auto rplus_parser = "+>" >> skp >> algebra_operand_parser ->* [](type t) {
+    return record_field{"+>", std::move(t)};
+  };
+  auto plus_parser = '+' >> skp >> algebra_operand_parser ->* [](type t) {
+    return record_field{"+", std::move(t)};
+  };
+  auto lplus_parser = "<+" >> skp >> algebra_operand_parser ->* [](type t) {
+    return record_field{"<+", std::move(t)};
+  };
+  auto minus_parser = '-' >> skp >> (parsers::identifier | parsers::qqstr) ->*
+    [](std::string field) {
+      record_type result;
+      result.fields.emplace_back(std::move(field), bool_type{});
+    return record_field{"-", std::move(result)};
+  };
+  auto algebra_parser
+    = rplus_parser
+    | plus_parser
+    | lplus_parser
+    | minus_parser
+    ;
+  type_expr_parser = (algebra_operand_parser >> skp >> (+(skp >> algebra_parser)))
+    ->* [](std::tuple<type, std::vector<record_field>> xs) -> type {
+      auto& [lhs, rhss] = xs;
+      record_type result;
+      result.fields = {record_field{"", std::move(lhs)}};
+      result.fields.insert(result.fields.end(), rhss.begin(), rhss.end());
+      return result.attributes({{"vast-algebra"}});
```

Comment:
Use a character that is not part of `identifier_chars` so this can not be constructed manually (`$algebra`).